Expose Split-DWARF support in Frontend and Driver

hjyamauchi opened the pull request:
This patch wires up flags in Driver and Frontend that expose support for GNU debug fission / DWARF5 standardized Split-DWARF as implemented in LLVM. It splits the majority of debug information from .o outputs into separate .dwo files.

This is based on https://github.com/apple/swift/pull/70296 authored by weliveindetail and adds a rebase and fix for the macos test failure.

hjyamauchi commented:
@swift-ci please test

hjyamauchi commented:
@swift-ci Please test Windows platform

compnerd commented on include/swift/AST/IRGenOptions.h:
```diff
@@ -239,6 +239,9 @@ class IRGenOptions {
   uint8_t DWARFVersion = 4;
 
+  /// Enables DWARF Fission and sets the file path for .dwo output.
+  std::string SplitDwarfOutput;
```
Perhaps `DWOPath` or `DWOOutput` is more appropriate of a name?

hjyamauchi replied:
Note Stefan's notes about the names [here](https://github.com/apple/swift/pull/70296#issuecomment-1857986159). I don't have preferences. Will update per suggestions.

adrian-prantl replied:
LLVM calls it SplitDebugFileName (https://llvm.org/doxygen/classllvm_1_1DICompileUnit.html#af664fcb825cb6ffc1567fc2bc244173f). Personally, as long as we have a "Path" or "FileName" suffix I'm fine with either DWO or Split as the prefix.

hjyamauchi replied:
Done

compnerd commented on include/swift/Option/Options.td:
```diff
@@ -1037,6 +1037,13 @@ def dwarf_version : Joined<["-"], "dwarf-version=">,
+def dwarf_fission : Flag<["-"], "dwarf-fission">,
+  Flags<[FrontendOption]>,
+  HelpText<"Emit debug info as split-DWARF (Fission) in a separate .dwo file">;
+def emit_split_dwarf_path : Separate<["-"], "split-dwarf-output">,
```
I think that `-dwarf-fission-path` would be a better option name. If you look at existing options, they match: `-emit-module -emit-module-path`.

hjyamauchi replied:
Done

compnerd commented:
Could you share the changes for apple/swift-driver as well please? I think that we should do those simultaneously as we don't ship the C++ driver on Windows at least.

adrian-prantl commented on include/swift/Driver/Driver.h:
```diff
@@ -128,6 +128,9 @@ class OutputInfo {
   std::optional<uint8_t> DWARFVersion;
 
+  /// Emit debug info as split-DWARF (Fission) in separate .dwo files.
+  bool DWARFFission = false;
```
Let's stick with either calling everything "Split" or "DWO". I think post DWARF 5, the "fission" name is less often used and could confuse people.

hjyamauchi replied:
OK. I will remove the "fission" in the names. @compnerd Please let me know if you are not on the same page.

compnerd replied:
Sure, I think that `SplitDWARF` is fine to use as an alternative to `DWARFFission`. We should be able to use the former throughout.

hjyamauchi replied:
Done

hjyamauchi committed:
Expose Split-DWARF support in Frontend and Driver

This patch wires up flags in Driver and Frontend that expose support
for GNU debug fission / DWARF5 standardized Split-DWARF as implemented
in LLVM. It splits the majority of debug information from .o outputs
into separate .dwo files.

This is based on #70296 authored by weliveindetail and adds a rebase,
fix for the macos test failure and some flag renaming.

hjyamauchi commented:
> Could you share the changes for apple/swift-driver as well please? I think that we should do those simultaneously as we don't ship the C++ driver on Windows at least.

Will do

adrian-prantl commented on include/swift/AST/IRGenOptions.h:
```diff
@@ -239,6 +239,9 @@ class IRGenOptions {
   uint8_t DWARFVersion = 4;
 
+  /// Enables Split DWARF and sets the file path for .dwo output.
+  std::string DWOPath;
```
How is this going to work in a multi-output compilation that produces more than one .o file? Is each instance of IRGen only responsible for a single .o file?

compnerd replied:
I believe that is the correct model - each IRGen ultimately creates a `.o` and when it does so, should generate its corresponding `.dwo`.
